Source/Windows/WOW64/IntervalList.h

```diff
@@ -0,0 +1,143 @@
+#pragma once
+
+#include <utility>
+#include <algorithm>
+
+#include <FEXCore/fextl/vector.h>
+
+template<typename SizeType>
+class IntervalList {
+public:
+  using DifferenceType = decltype(std::declval<SizeType>() - std::declval<SizeType>());
+
+  struct Interval {
+    SizeType Offset;
+    SizeType End;
+
+    Interval() = default;
+
+    Interval(SizeType Offset, SizeType End) : Offset{Offset}, End{End} {}
+  };
+
+private:
+  fextl::vector<Interval> Intervals; ///< list of intervals sorted by their end offset
+
+public:
+  struct QueryResult {
+    bool Enclosed; ///< If the given offset was enclosed by an interval
+    DifferenceType Size; ///< Size of the interval starting from the query offset, or distance to the next interval if
+                         /// `Enclosed` is false (if there is no next interval, size is 0)
+  };
+
+  void Clear() {
+    Intervals.clear();
+  }
+
+  void Insert(Interval Entry) {
+    auto FirstIt = std::upper_bound(Intervals.begin(), Intervals.end(), Entry, [](const auto &LHS, const auto &RHS) {
+      return LHS.Offset < RHS.End;
+    }); // Lowest offset interval that (maybe) overlaps with the new interval
+
+    if (FirstIt == Intervals.end() || FirstIt->Offset >= Entry.End) {
+      Intervals.insert(FirstIt, Entry);
+      return;
+    }
+    // Now FirstIt will always overlap
+
+    auto LastIt = FirstIt; // Highest offset interval that overlaps with the new interval
+    while ((std::next(LastIt) != Intervals.end()) && (std::next(LastIt)->Offset < Entry.End)) {
```

Comment:
(superseded by equal_range now but) why would the loop terminate as you say? It would incr lastIt till the element after it has an offset >= to the entries end, but in this case it wouldn't hit that and would end being prev(Intervals.end()) as expected